src/main/java/com/puppycrawl/tools/checkstyle/SarifLogger.java

```diff
@@ -0,0 +1,348 @@
+    private final boolean closeStream;
+
+    /** The results. */
+    private final List<String> results = new ArrayList<>();
+
+    /** Content for the entire report. */
+    private final String report;
+
+    /** Content for result representing an error with source line and column. */
+    private final String resultLineColumn;
+
+    /** Content for result representing an error with source line only. */
+    private final String resultLineOnly;
+
+    /** Content for result representing an error with filename only and without source location. */
+    private final String resultFileOnly;
+
+    /** Content for result representing an error without filename or location. */
+    private final String resultErrorOnly;
+
+    /**
+     * Creates a new {@code SarifLogger} instance.
+     *
+     * @param outputStream where to log audit events
+     * @param outputStreamOptions if {@code CLOSE} that should be closed in auditFinished()
+     * @throws IllegalArgumentException if outputStreamOptions is null
+     * @throws IOException if there is reading errors.
+     */
+    public SarifLogger(
+        OutputStream outputStream,
+        OutputStreamOptions outputStreamOptions) throws IOException {
+        writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+        if (outputStreamOptions == null) {
+            throw new IllegalArgumentException("Parameter outputStreamOptions can not be null");
+        }
+        closeStream = outputStreamOptions == OutputStreamOptions.CLOSE;
+        report = readResource("/com/puppycrawl/tools/checkstyle/sarif/SarifReport.template");
+        resultLineColumn =
+            readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultLineColumn.template");
+        resultLineOnly =
+            readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultLineOnly.template");
+        resultFileOnly =
+            readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultFileOnly.template");
+        resultErrorOnly =
+            readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultErrorOnly.template");
+    }
+
+    @Override
+    protected void finishLocalSetup() {
+        // No code by default
+    }
+
+    @Override
+    public void auditStarted(AuditEvent event) {
+        // No code by default
+    }
+
+    /**
+     * {@inheritDoc}
+     * Following idea suppressions are false positives
+     *
+     * @noinspection DynamicRegexReplaceableByCompiledPattern
+     */
+    @Override
+    public void auditFinished(AuditEvent event) {
+        final String version = SarifLogger.class.getPackage().getImplementationVersion();
+        final StringBuilder renderedResults = new StringBuilder(results.size());
+        for (int i = 0; i < results.size(); ++i) {
+            if (i > 0) {
+                renderedResults.append(',');
+            }
+            renderedResults.append('\n');
+            renderedResults.append(results.get(i));
+        }
```

Comment:
it is easier to make it `String.join(",\n", results)` instead of for-cycle

Reply:
👍  I didn't know this method exists.